compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/SignedRemNode.java

```diff
@@ -81,27 +85,55 @@ private static ValueNode canonical(SignedRemNode self, ValueNode forX, ValueNode
             IntegerStamp yStamp = (IntegerStamp) forY.stamp(view);
             if (constY < 0 && constY != CodeUtil.minValue(yStamp.getBits())) {
                 Stamp newStamp = IntegerStamp.OPS.getRem().foldStamp(forX.stamp(view), forY.stamp(view));
-                return canonical(null, forX, ConstantNode.forIntegerStamp(yStamp, -constY), zeroCheck, newStamp, view);
+                return canonical(null, forX, ConstantNode.forIntegerStamp(yStamp, -constY), zeroCheck, newStamp, view, tool);
             }
 
             if (constY == 1) {
                 return ConstantNode.forIntegerStamp(stamp, 0);
-            } else if (CodeUtil.isPowerOf2(constY)) {
-                if (xStamp.isPositive()) {
-                    // x & (y - 1)
-                    return new AndNode(forX, ConstantNode.forIntegerStamp(stamp, constY - 1));
-                } else if (xStamp.isNegative()) {
-                    // -((-x) & (y - 1))
-                    return new NegateNode(new AndNode(new NegateNode(forX), ConstantNode.forIntegerStamp(stamp, constY - 1)));
+            } else if (CodeUtil.isPowerOf2(constY) && tool != null && tool.allUsagesAvailable()) {
+                if (allUsagesCompareAgainstZero(self)) {
```

Comment:
@XiaohongGong when all usages just compare against zero, we can transform this to `(x&(y-1))==0`. Then we do not need the complex match rule because the case `x&1==0` covers it.

Reply:
Here I created a [proof](https://rise4fun.com/Z3/XPZk) of `((x%n)==0)==((x&(((1<<n)-1))==0)`:
```
(set-logic QF_BV)
; Constants
(declare-const one (_ BitVec 32))
(assert (= one (_ bv1 32)))
(declare-const zero (_ BitVec 32))
(assert (= zero (_ bv0 32)))

; Inputs
(declare-const x (_ BitVec 32))
(declare-const y (_ BitVec 32))
(declare-const n (_ BitVec 32))              ; nth power
(assert (= y (bvshl one n)))                 ; y = 1 << n;

; reference implementation
(declare-const reference Bool)
(assert (= reference (= (bvsrem x y) zero)))     ; reference = (x%y)==0

; optimization
(declare-const opt Bool)
(assert (= opt (= (bvand x (bvsub y one)) zero))) ; opt = (x&(y-1))==0

; proof reference=opt
(assert (not (= reference opt))) ; check whether reference!=opt is satisfiable; if unsatisfiable the opposite must be true
(check-sat)
```

Reply:
Wonderful!